core-skills/web-application-development-with-react/fundamentals/README.md

```diff
@@ -0,0 +1,18 @@
+# React Fundamentals
+
+## Summary
+
+React is a framework with a number of moving parts, components, states, props, etc. It
+can be very easy to begin writing components without properly understanding how each of
+these should be used and how to use them.
+
+An Engineer who has passed this core skill would be expected to understand what each of these
+parts are, and when to use one over the other (i.e. When should you put something in a prop instead
+of the state).
+
+The first two sections of React's documentation, [JSX](https://reactjs.org/docs/introducing-jsx.html) and [Rendering](https://reactjs.org/docs/rendering-elements.html) are important reading.
+
+- [Components and Props](https://reactjs.org/docs/components-and-props.html)
```

Comment:
Do we want to include context in the fundamentals?

Reply:
Maybe? I was debating it - but i'm not sure whether or not it is a fundamental or a more advanced concept. You could compare it to using component composition potentially? But i'm not sure what level that would be

Reply:
Having looked into it I'm erring on the side of no? I it should be mentioned in Fox potentially, but you can get by without it so i wouldn't class it as a 'Fundamental'

Reply:
I'd argue the context APIs are important for any reasonably involved React app, would personally like to see them included. Given there's a good chance engineers will eventually have to use it on client projects (or should use it without knowing it) and there's a 100% chance they'll use it in libraries (e.g. redux, react router, theme providers).

Think they're often seen as dark magic, so hopefully a React course like this would be able to properly explain it.

Reply:
:+1: I will defer to you on this one as I've got less experience with them which has made me think they might be less important

Reply:
Wouldn't that be more appropriate for Fox or Owl?

Reply:
So imo for contexts:

- It's used in a lot of tools, so it's pretty fundamental knowledge to get how they work/how you'd test them in future

But 

- It's not required for building a very basic react app 

So i'm not sure whether or not I'd consider it fundamental or not

Reply:
I'd consider it Fox and above in that case

Reply:
What about https://reactjs.org/docs/refs-and-the-dom.html refs

Reply:
> "In the typical React dataflow, props are the only way that parent components interact with their children... However, there are a few cases where you need to imperatively modify a child outside of the typical dataflow "

This doesn't sound like a fundamental to me. By the end of this core skill (at least, the first three levels if more got added) - I would expect the person to be empowered enough to be able to look up things like references and understand what they do because they have a good understanding of the React fundamentals.

We could very easily fall into the trap of believing everything is a fundamental - but we should make sure what we mark down actually _are_ fundamental to Rect knowledge. Arguably Context isn't due to you needing to have an understanding of what props are to begin with, but a lot of tools use context which makes me think it might be worthwhile